File: apps/ios/Linkfit/Core/Analytics/Analytics.swift
import Foundation
#if canImport(PostHog)
import PostHog
#endif
import os.log

// =============================================================================
// ANALYTICS FACADE
// -----------------------------------------------------------------------------
// Thin call-site facade over the PostHog iOS SDK. Three contracts the rest of
// the app depends on:
//
//   1. **Safe with no API key.** `start(apiKey:host:)` with a nil/empty key
//      is a no-op — every subsequent call short-circuits. This lets dev/CI
//      builds boot without configuring PostHog. The same is true when the
//      `PostHog` module isn't linked at all (the package is opt-in via
//      SwiftPM and we guard with `canImport`).
//
//   2. **Typed events only.** `track(_:)` accepts an `AnalyticsEvent` enum
//      so the schema lives in one place and call sites can't drift. A
//      legacy free-form `track(_:_:)` is kept for the in-house buffer (see
//      `AnalyticsClient.swift`) so existing call sites don't break — we
//      forward those into PostHog too when the SDK is configured.
//
//   3. **PII boundary.** `identify(userId:traits:)` only sends the
//      server-side UUID and explicitly enumerated traits. The call site
//      MUST NOT pass email or phone — strip those at the boundary. The
//      backend hashes emails before they reach PostHog; the iOS layer
//      simply never sees the raw value.
//
// The release version is read from the bundle; the host is read from the
// `POSTHOG_HOST` Info.plist value with `https://app.posthog.com` as the
// fallback. Production builds substitute both keys at xcodegen time from
// the `$(POSTHOG_API_KEY)` / `$(POSTHOG_HOST)` env vars — see
// `project.yml`.
// =============================================================================

/// Call-site entry point for analytics. All methods are no-ops when the SDK
/// has not been configured with a non-empty API key, so callers don't need
/// to special-case dev/CI builds.
enum Analytics {
    private static let log = Logger(subsystem: "az.linkfit.app", category: "Analytics")

    /// Sentinel flipped after the first successful `start(...)` call.
    /// Subsequent `track`/`identify`/`reset` calls short-circuit when this
    /// stays false (no API key or SDK not linked).
    nonisolated(unsafe) private static var isStarted: Bool = false

    // MARK: - Bootstrap

    /// Initialize the PostHog client. Idempotent — a second call with the
    /// same key returns immediately. A nil or empty `apiKey` makes every
    /// subsequent call a no-op, so this is safe to invoke from
    /// `LinkfitApp.init` without a configured project.
    ///
    /// - Parameters:
    ///   - apiKey: PostHog project API key. Read from the `POSTHOG_API_KEY`
    ///     Info.plist value at the call site. Nil/empty → no-op.
    ///   - host: PostHog cloud host. Read from `POSTHOG_HOST` with
    ///     `https://app.posthog.com` as the default.
    static func start(apiKey: String?, host: String?) {
        guard !isStarted else { return }
        let trimmedKey = apiKey?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !trimmedKey.isEmpty else {
            log.info("PostHog disabled — no API key configured")
            return
        }

        #if canImport(PostHog)
        let resolvedHost: String = {
            let trimmedHost = host?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            return trimmedHost.isEmpty ? "https://app.posthog.com" : trimmedHost
        }()

        // PostHog 3.x renamed `apiKey` → `projectToken` for the init; the
        // overload with both fields takes `(projectToken:host:)`. We pass
        // the trimmed key + resolved host so a custom EU / self-hosted
        // instance can be wired via `POSTHOG_HOST` without code changes.
        let config = PostHogConfig(projectToken: trimmedKey, host: resolvedHost)
        // Sessions + session replay are off by default — turn on later
        // when we're ready to deal with the privacy review. Autocapture
        // is left at the SDK default (off on iOS today) for the same
        // reason.
        config.captureScreenViews = false
        config.captureApplicationLifecycleEvents = true
        // We control flushes ourselves on background/foreground via the
        // SDK's default batching policy (every 30s / 20 events). No need
        // to override here.
        PostHogSDK.shared.setup(config)
        isStarted = true
        log.info("PostHog initialized")
        #else
        // Package not linked — leave `isStarted` false so every subsequent
        // call is a clean no-op. Don't log an error: the brief explicitly
        // permits this build configuration in dev/CI.
        _ = trimmedKey
        #endif
    }

    // MARK: - Identification

    /// Tag subsequent events with a stable user identifier. Call this once
    /// after sign-in or registration succeeds; the `userId` is the
    /// server-side UUID, never an email or phone number.
    ///
    /// Traits map onto PostHog person properties — keep them to a small,
    /// auditable set (locale, signup_method, etc.). NEVER pass `email` or
    /// `phone` here — the backend is responsible for hashed email
    /// properties; the iOS surface never sees raw values.
    static func identify(userId: String, traits: [String: Any] = [:]) {
        let cleanedId = userId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !cleanedId.isEmpty else { return }

        #if canImport(PostHog)
        if isStarted {
            if traits.isEmpty {
                PostHogSDK.shared.identify(cleanedId)
            } else {
                PostHogSDK.shared.identify(cleanedId, userProperties: traits)
            }
        }
        #endif

        // Mirror the distinct_id into the in-house buffer so events
        // queued by `AnalyticsClient` carry the same identifier even
        // when PostHog is not configured (dev/CI builds still want
        // events tagged with the right user id for the server-side
        // `/api/v1/analytics/events` consumer).
        AnalyticsClient.shared.assumeIdentify(cleanedId)
    }

    // MARK: - Tracking — typed

    /// Emit a typed event. The enum case enforces the property schema so
    /// the dashboard never sees a malformed payload from this layer.
    ///
    /// Also mirrors into the legacy in-house buffer so server-side
    /// analytics (which currently consumes `/api/v1/analytics/events`)
    /// continue to receive the same stream — once we're confident in
    /// PostHog as the single source of truth this mirror can drop.
    static func track(_ event: AnalyticsEvent) {
        let name = event.name
        let properties = event.properties

        #if canImport(PostHog)
        if isStarted {
            PostHogSDK.shared.capture(name, properties: properties)
        }
        #endif

        // Mirror into the in-house buffer using the typed-value variant.
        // Translate the heterogeneous `[String: Any]` into
        // `[String: AnalyticsValue]` — only scalar values pass through
        // (matching `AnalyticsValue`'s closed enum); anything else is
        // dropped silently. The enum's own contract guarantees scalars,
        // so the drop branch is dead code in practice.
        Analytics.track(name, Self.toAnalyticsValues(properties))
    }

    // MARK: - Tracking — legacy free-form

    /// Buffer an event for later flush via the in-house pipeline. Kept for
    /// backward compatibility with existing call sites that wrote
    /// `Analytics.track("foo", [...])` before the typed event enum
    /// landed. Prefer `track(_ event: AnalyticsEvent)` for new code.
    static func track(_ event: String, _ properties: [String: AnalyticsValue] = [:]) {
        // Mirror to PostHog — convert AnalyticsValue back to scalar Any
        // so the vendor SDK doesn't need to know our enum.
        #if canImport(PostHog)
        if isStarted {
            let flat = Self.toAnyDict(properties)
            PostHogSDK.shared.capture(event, properties: flat)
        }
        #endif

        Task { await AnalyticsClient.shared.track(event, properties: properties) }
    }

    /// Tag subsequent events with a stable user identifier. Call this once
    /// after sign-in succeeds. Legacy overload — new code should use the
    /// `identify(userId:traits:)` variant.
    static func identify(_ userId: String) {
        identify(userId: userId, traits: [:])
    }

    /// Drop the distinct_id and the pending buffer. Call this on logout
    /// so the next user's session can't be attributed to the previous
    /// one. Resets PostHog's distinct_id when the SDK is configured.
    static func reset() {
        #if canImport(PostHog)
        if isStarted {
            PostHogSDK.shared.reset()
        }
        #endif
        Task { await AnalyticsClient.shared.reset() }
    }

    /// Force a flush. Useful right before a process terminates (e.g.
    /// inside a background-task expiration handler) or in tests.
    static func flush() {
        #if canImport(PostHog)
        if isStarted {
            PostHogSDK.shared.flush()
        }
        #endif
        Task { await AnalyticsClient.shared.flush() }
    }

    // MARK: - Private conversion helpers

    /// Translate a property bag of scalars (`String`/`Int`/`Bool`/
    /// `Double`) into the closed `AnalyticsValue` enum used by the
    /// in-house buffer. Non-scalar values are silently dropped — the
    /// typed-event enum guarantees scalars at the source, so this branch
    /// is dead in practice and exists only as defense-in-depth.
    private static func toAnalyticsValues(_ raw: [String: Any]) -> [String: AnalyticsValue] {
        var out: [String: AnalyticsValue] = [:]
        out.reserveCapacity(raw.count)
        for (key, value) in raw {
            // Bool MUST be checked before Int — on iOS, `Bool` bridges
            // to `NSNumber` which also matches `Int`. Without this order
            // a `true` would land as `1`.
            if let bool = value as? Bool {
                out[key] = .bool(bool)
            } else if let int = value as? Int {
                out[key] = .int(int)
            } else if let double = value as? Double {
                out[key] = .double(double)
            } else if let string = value as? String {
                out[key] = .string(string)
            }
            // Else: drop. The typed-event enum never produces non-scalars.
        }
        return out
    }

    /// Convert the legacy `AnalyticsValue` bag into the heterogeneous
    /// dictionary PostHog's `capture` method expects.
    private static func toAnyDict(_ values: [String: AnalyticsValue]) -> [String: Any] {
        var out: [String: Any] = [:]
        out.reserveCapacity(values.count)
        for (key, value) in values {
            switch value {
            case .string(let s): out[key] = s
            case .int(let i):    out[key] = i
            case .double(let d): out[key] = d
            case .bool(let b):   out[key] = b
            }
        }
        return out
    }
}

// =============================================================================
// IN-HOUSE BUFFER BRIDGE
// -----------------------------------------------------------------------------
// `AnalyticsClient` exposes async `identify(_:)` already, but we want a
// fire-and-forget surface so the PostHog facade can mirror the distinct_id
// without awaiting. Extension lives in this file (rather than in
// AnalyticsClient.swift) so the legacy file is untouched by this agent.
// =============================================================================

extension AnalyticsClient {
    /// Fire-and-forget identity sync. Called by the PostHog facade when
    /// `Analytics.identify(userId:traits:)` runs so the in-house buffer
    /// tags subsequent events with the same id.
    nonisolated func assumeIdentify(_ userId: String) {
        Task { [weak self] in await self?.identify(userId) }
    }
}

File: apps/api/src/shared/observability/analytics.ts
/**
 * Server-side product analytics facade.
 *
 * Wraps `posthog-node` behind a tiny surface so the rest of the API never
 * imports the vendor SDK directly. Three contracts the rest of the codebase
 * relies on:
 *
 * 1. **Safe with no API key.** `initAnalytics({ apiKey: undefined | "" })` is
 *    a no-op — every subsequent `track(...)` call short-circuits. This lets
 *    dev/CI environments boot without configuring PostHog.
 * 2. **PII boundary.** Email and phone are NEVER sent in cleartext. The
 *    {@link hashEmail} helper produces a stable SHA-256 of the lowercased
 *    + trimmed email so funnels can be joined to other systems without
 *    leaking the raw identifier. Call sites are responsible for hashing
 *    BEFORE handing the value to `track(...)` — this module does not
 *    rummage through `properties` looking for emails.
 * 3. **Fire-and-forget.** `track(...)` never throws and never blocks the
 *    request hot path. Failures are swallowed with a single `console.warn`.
 *    Backpressure is handled by the SDK's internal batching (every 20
 *    events or 30 seconds, whichever first).
 *
 * Server-side events fire at three call sites today (see the comment block
 * at the bottom of this file for the taxonomy):
 *   - `account_created` — in the auth signup handler.
 *   - `game_created`    — in `games.service.ts` `create()`.
 *   - `match_scored_finalized` — in `scoring.service.ts` `complete()`.
 *
 * Each call site looks up the distinct_id (the user UUID) and passes the
 * minimum set of properties needed to draw the funnel. Avoid free-form
 * properties — they make PostHog dashboards painful to query and the
 * audit surface for PII regressions explodes.
 */
import { PostHog } from "posthog-node";
import crypto from "node:crypto";

/** Sentinel that flips true on the first successful `initAnalytics(...)`. */
let client: PostHog | null = null;

export interface InitAnalyticsOptions {
  /**
   * PostHog project API key. Undefined or empty disables the integration
   * entirely. The Linkfit dev/CI environment leaves this unset; production
   * sets it via `POSTHOG_API_KEY` (see `shared/config/env.ts`).
   */
  apiKey: string | undefined;
  /**
   * PostHog host. Defaults to `https://app.posthog.com` (PostHog Cloud).
   * EU customers should point at `https://eu.posthog.com`; self-hosted
   * deployments at their own URL.
   */
  host?: string;
  /**
   * Optional logger for init/teardown notices. Defaults to no-op so the
   * module stays usable from tests without wiring a pino instance.
   */
  logger?: { info: (msg: string) => void; warn: (msg: string) => void };
}

/**
 * Initialize the PostHog client. Idempotent — a second call with the same
 * key returns immediately. Passing an empty / undefined `apiKey` keeps the
 * module in no-op mode for the lifetime of the process.
 */
export function initAnalytics(options: InitAnalyticsOptions): void {
  if (client !== null) {
    return;
  }
  const apiKey = options.apiKey?.trim();
  if (apiKey === undefined || apiKey.length === 0) {
    options.logger?.info("PostHog disabled — POSTHOG_API_KEY not set");
    return;
  }
  const trimmedHost = options.host?.trim() ?? "";
  const host = trimmedHost.length > 0 ? trimmedHost : "https://app.posthog.com";
  client = new PostHog(apiKey, {
    host,
    // Batch settings — match the SDK defaults but pin them here so a
    // future minor version can't silently change throughput.
    flushAt: 20,
    flushInterval: 30_000,
  });
  options.logger?.info("PostHog initialized");
}

/**
 * Shut down the PostHog client and flush any pending events. Call this
 * from the server shutdown handler so events queued in the last 30s are
 * not lost when the process exits.
 *
 * Safe to call when the client was never initialized — returns immediately.
 */
export async function shutdownAnalytics(): Promise<void> {
  if (client === null) {
    return;
  }
  try {
    await client.shutdown();
  } catch {
    // Swallow — we're shutting down anyway.
  }
  client = null;
}

export interface TrackInput {
  /**
   * Stable user id. For authenticated events this is the user UUID; for
   * pre-auth events (e.g. waitlist signup) it is the email hash. Required
   * — PostHog rejects events with an empty distinct_id.
   */
  distinctId: string;
  /** Event name in snake_case. Stable wire identifier. */
  event: string;
  /** Property bag. Scalars only (string/number/boolean). NO emails or phones. */
  properties?: Record<string, string | number | boolean | null | undefined>;
}

/**
 * Buffer an event for later flush. Non-throwing — failures are swallowed
 * with a single `console.warn`. No-op when the client was never
 * initialized (dev/CI without a configured PostHog project).
 *
 * The call site MUST NOT pass raw email or phone in `properties`. Use
 * {@link hashEmail} to derive an `email_hash` field instead.
 */
export function track(input: TrackInput): void {
  if (client === null) {
    return;
  }
  if (input.distinctId.length === 0) {
    return;
  }
  try {
    // `exactOptionalPropertyTypes` rejects `properties: undefined`; PostHog
    // expects either a populated record or the key to be absent. Defaulting
    // to `{}` keeps the call site signature optional while satisfying the
    // SDK's typed interface.
    client.capture({
      distinctId: input.distinctId,
      event: input.event,
      properties: input.properties ?? {},
    });
  } catch (error) {
    // Vendor SDK errors are never allowed to break the request path.
    console.warn("analytics capture failed:", error);
  }
}

/**
 * Associate user properties with a distinct id. Use this once on
 * registration to seed PostHog's person profile with locale, signup
 * method, and other audit-safe traits. NEVER pass email or phone in
 * `traits` — hash them through {@link hashEmail} first if needed.
 */
export function identify(
  distinctId: string,
  traits: Record<string, string | number | boolean | null | undefined> = {},
): void {
  if (client === null || distinctId.length === 0) {
    return;
  }
  try {
    client.identify({ distinctId, properties: traits });
  } catch (error) {
    console.warn("analytics identify failed:", error);
  }
}

/**
 * Stable SHA-256 hash of an email address, suitable as an `email_hash`
 * property in analytics events. Lowercases and trims first so casing /
 * whitespace differences across signup methods (manual, Apple, Google)
 * produce the same hash.
 *
 * Returns the hex digest. Empty input returns an empty string so callers
 * can pass it through without an extra null-check (the consumer's PostHog
 * column will be NULL rather than the hash of an empty string).
 */
export function hashEmail(email: string | null | undefined): string {
  if (email === undefined || email === null) {
    return "";
  }
  const normalized = email.trim().toLowerCase();
  if (normalized.length === 0) {
    return "";
  }
  return crypto.createHash("sha256").update(normalized).digest("hex");
}

/**
 * Extract the domain part of an email for low-cardinality bucketing in
 * dashboards. Returns an empty string for malformed inputs. Lowercased
 * so `Gmail.COM` and `gmail.com` bucket together.
 *
 * This is intentionally lossy — the domain is fine to ship in cleartext
 * since it identifies an email provider, not an individual user.
 */
export function emailDomain(email: string | null | undefined): string {
  if (email === undefined || email === null) {
    return "";
  }
  const trimmed = email.trim().toLowerCase();
  const at = trimmed.lastIndexOf("@");
  if (at === -1 || at === trimmed.length - 1) {
    return "";
  }
  return trimmed.slice(at + 1);
}

/** Returns true when initAnalytics has run successfully with a non-empty key. */
export function isAnalyticsEnabled(): boolean {
  return client !== null;
}

/**
 * Test-only reset. Vitest spins multiple `buildServer` calls per file and
 * the client is a process-wide singleton — call this in test setup so a
 * fresh `initAnalytics` per test is possible. NOT for production use.
 */
export function _resetForTesting(): void {
  client = null;
}

// ─────────────────────────────────────────────────────────────────────────
// Event taxonomy — server-side
// ─────────────────────────────────────────────────────────────────────────
//
// The wave-10 launch ships three server-side events. Keep this list in
// sync with the iOS taxonomy in `Core/Analytics/AnalyticsEvent.swift` and
// the PostHog dashboard's "events" reference page.
//
//   account_created          { email_domain, locale, email_hash }
//   game_created             { visibility, capacity, sport_slug? }
//   match_scored_finalized   { duration_min, winner_team?, sets_played? }
//
// Each event MUST carry the actor's user UUID as `distinctId`. The
// `email_hash` property is optional on `account_created` and intentionally
// omitted on the other two (we only need the join key at signup time —
// later events tie themselves to the same distinct_id PostHog already
// resolved).
